python/singa/autograd.py

```diff
@@ -2760,3 +2760,42 @@ def backward(self, dy):
 
 def reciprocal(x):
     return Reciprocal()(x)[0]
+
+
+class GlobalAveragePool(Operation):
+    def __init__(self, data_format='channels_first'):
+        super(GlobalAveragePool, self).__init__()
+        self.data_format = data_format
+
+    def forward(self, x):
+        if training:
+            self.mask = singa.Tensor(list(x.shape()), x.device())
+
+        shape = list(x.shape())
+        
+        # (N x C x H x W) for channels_first
+        if self.data_format == 'channels_first':
+            axes = tuple(i for i in range(2, len(shape)))
+            self.shape_divisor = 1/np.prod(shape[2:])
+        else: # (N x H x W x C) for channels_last
+            axes = tuple(i for i in range(1, len(shape)-1))
+            self.shape_divisor = 1/np.prod(shape[1:-1])
+
+        # output shape
+        # (N x C x 1 x 1) for channels_first
+        # (N x 1 x 1 x C) for channels_last
+        for i in axes:
+            shape[i] = 1
+
+        x = tensor.from_raw_tensor(x)
+        x = tensor.sum(x, axis=axes)
```

Comment:
the raw tensor should also have the sum function?

Reply:
it seems the singa.Sum does not support sum over axis larger than 1.